### ISSUE TITLE
Ensure that updates don't start breaking macOS builds in Core

### ISSUE THREAD

**Issue by dmsnell** [WordPress/gutenberg#57653]
### Description

Recently an update here in Gutenberg started causing Core builds to start failing on the macOS runner. The problem was that the `npm run build` step failed with Node running out of memory.

@desrosj fixed this by upping the Node heap size in the Core test runners, but it would be better if we could flag a PR that triggers a failure of any kind so that we don't have to start debugging unexpected failures (because these are not related to changes in Core, but rather to changes in Gutenberg).

Can we start building Gutenberg in this repo in macOS with a setup similar to what's on the Core side?

### Step-by-step reproduction instructions

Watch the Core repo and wait for failures, they will come unexpectedly, just as changes in Core will unexpectedly break Gutenberg.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by dmsnell** [comment]
Nothing was done in response to this, so I assume it's still a problem. Here's to neglect and hoping 🍻

**Comment by Successfulsebunya** [comment]
## Reproduction Report
### Description
This report validates whether the issue can be reproduced.

### Environment
- WordPress: 6.8.2
- PHP: 8.2.29
- Server: Apache/2.4.62 (Debian)
- Database: mysqli (Server: 11.8.2-MariaDB-ubu2404 / Client: mysqlnd 8.2.29)
- Browser: Firefox 141.0
- OS: macOS
- Theme: Twenty Twenty-Three 1.6
- MU Plugins: None activated
- Plugins:
  * My Test Plugin 0.1.0
  * Test Reports 1.2.0

### Actual Results
1.  - ❌ Error condition does not occur (not reproduced).

### Additional Notes
- No memory errors found. Problem is solved

### Supplemental Artifacts
mocursor@MoCursor-MBP gutenberg % npm install --force

npm warn using --force Recommended protections disabled.
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: ajv-keywords@3.5.2
npm warn Found: ajv@8.7.1
npm warn node_modules/ajv
npm warn   dev ajv@"8.7.1" from the root project
npm warn   4 more (@sidvind/better-ajv-errors, ajv-errors, ajv-formats, table)
npm warn
npm warn Could not resolve dependency:
npm warn peer ajv@"^6.9.1" from ajv-keywords@3.5.2
npm warn node_modules/ajv-keywords
npm warn   ajv-keywords@"^3.5.2" from schema-utils@3.3.0
npm warn   node_modules/@pmmmwh/react-refresh-webpack-plugin/node_modules/schema-utils
npm warn   7 more (webpack, schema-utils, schema-utils, schema-utils, ...)
npm warn
npm warn Conflicting peer dependency: ajv@6.12.6
npm warn node_modules/ajv
npm warn   peer ajv@"^6.9.1" from ajv-keywords@3.5.2
npm warn   node_modules/ajv-keywords
npm warn     ajv-keywords@"^3.5.2" from schema-utils@3.3.0
npm warn     node_modules/@pmmmwh/react-refresh-webpack-plugin/node_modules/schema-utils
npm warn     7 more (webpack, schema-utils, schema-utils, schema-utils, ...)
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: react-native@0.73.3
npm warn Found: react@18.3.1
npm warn node_modules/react
npm warn   dev react@"18.3.1" from the root project
npm warn   113 more (@ariakit/react, @ariakit/react-core, @ariakit/test, ...)
npm warn
npm warn Could not resolve dependency:
npm warn peer react@"18.2.0" from react-native@0.73.3
npm warn node_modules/react-native
npm warn   dev react-native@"0.73.3" from the root project
npm warn   25 more (@emotion/native, @react-native-clipboard/clipboard, ...)
npm warn
npm warn Conflicting peer dependency: react@18.2.0
npm warn node_modules/react
npm warn   peer react@"18.2.0" from react-native@0.73.3
npm warn   node_modules/react-native
npm warn     dev react-native@"0.73.3" from the root project
npm warn     25 more (@emotion/native, @react-native-clipboard/clipboard, ...)
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: storybook-source-link@2.0.9
npm warn Found: @storybook/components@8.4.7
npm warn node_modules/@storybook/components
npm warn   peer @storybook/components@"^8.3.0" from @geometricpanda/storybook-addon-badges@2.0.5
npm warn   node_modules/@geometricpanda/storybook-addon-badges
npm warn     dev @geometricpanda/storybook-addon-badges@"2.0.5" from the root project
npm warn   1 more (@storybook/react)
npm warn
npm warn Could not resolve dependency:
npm warn peer @storybook/components@"^6.4.0" from storybook-source-link@2.0.9
npm warn node_modules/storybook-source-link
npm warn   dev storybook-source-link@"2.0.9" from the root project
npm warn
npm warn Conflicting peer dependency: @storybook/components@6.5.16
npm warn node_modules/@storybook/components
npm warn   peer @storybook/components@"^6.4.0" from storybook-source-link@2.0.9
npm warn   node_modules/storybook-source-link
npm warn     dev storybook-source-link@"2.0.9" from the root project
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: storybook-source-link@2.0.9
npm warn Found: @storybook/core-events@8.6.14
npm warn node_modules/@storybook/core-events
npm warn   peer @storybook/core-events@"^8.3.0" from @geometricpanda/storybook-addon-badges@2.0.5
npm warn   node_modules/@geometricpanda/storybook-addon-badges
npm warn     dev @geometricpanda/storybook-addon-badges@"2.0.5" from the root project
npm warn
npm warn Could not resolve dependency:
npm warn peer @storybook/core-events@"^6.4.0" from storybook-source-link@2.0.9
npm warn node_modules/storybook-source-link
npm warn   dev storybook-source-link@"2.0.9" from the root project
npm warn
npm warn Conflicting peer dependency: @storybook/core-events@6.5.16
npm warn node_modules/@storybook/core-events
npm warn   peer @storybook/core-events@"^6.4.0" from storybook-source-link@2.0.9
npm warn   node_modules/storybook-source-link
npm warn     dev storybook-source-link@"2.0.9" from the root project
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: storybook-source-link@2.0.9
npm warn Found: @storybook/theming@8.4.7
npm warn node_modules/@storybook/theming
npm warn   dev @storybook/theming@"8.4.7" from the root project
npm warn   2 more (@geometricpanda/storybook-addon-badges, @storybook/react)
npm warn
npm warn Could not resolve dependency:
npm warn peer @storybook/theming@"^6.4.0" from storybook-source-link@2.0.9
npm warn node_modules/storybook-source-link
npm warn   dev storybook-source-link@"2.0.9" from the root project
npm warn
npm warn Conflicting peer dependency: @storybook/theming@6.5.16
npm warn node_modules/@storybook/theming
npm warn   peer @storybook/theming@"^6.4.0" from storybook-source-link@2.0.9
npm warn   node_modules/storybook-source-link
npm warn     dev storybook-source-link@"2.0.9" from the root project
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: typedoc@0.23.28
npm warn Found: typescript@5.7.2
npm warn node_modules/typescript
npm warn   dev typescript@"5.7.2" from the root project
npm warn   19 more (@appium/typedoc-plugin-appium, cosmiconfig, ...)
npm warn
npm warn Could not resolve dependency:
npm warn peer typescript@"4.6.x || 4.7.x || 4.8.x || 4.9.x || 5.0.x" from typedoc@0.23.28
npm warn node_modules/typedoc
npm warn   typedoc@"0.23.28" from @appium/docutils@0.4.10
npm warn   node_modules/@appium/docutils
npm warn   3 more (@appium/typedoc-plugin-appium, ...)
npm warn
npm warn Conflicting peer dependency: typescript@5.0.4
npm warn node_modules/typescript
npm warn   peer typescript@"4.6.x || 4.7.x || 4.8.x || 4.9.x || 5.0.x" from typedoc@0.23.28
npm warn   node_modules/typedoc
npm warn     typedoc@"0.23.28" from @appium/docutils@0.4.10
npm warn     node_modules/@appium/docutils
npm warn     3 more (@appium/typedoc-plugin-appium, ...)

> gutenberg@21.2.0 postinstall
> patch-package && node ./patches/patch-xcode.js

patch-package 8.0.0
Applying patches...
lighthouse@12.2.2 ✔
react-autosize-textarea@7.1.0 ✔
react-devtools-core@4.28.5 ✔
react-native@0.73.3 ✔
storybook-source-link@2.0.9 ✔
wasm-vips@0.0.10 ✔

> gutenberg@21.2.0 prepare
> husky install

fatal: not a git repository (or any of the parent directories): .git

added 37 packages, removed 1 package, changed 2 packages, and audited 4335 packages in 2m

446 packages are looking for funding
  run `npm fund` for details

65 vulnerabilities (7 low, 23 moderate, 32 high, 3 critical)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

**Comment by SirLouen** [comment]
### Additional Comments

> Step-by-step reproduction instructions
> Watch the Core repo and wait for failures, they will come unexpectedly, just as changes in Core will unexpectedly break Gutenberg. 

- @dmsnell can you provide more instructions on how to test this? It appears that this report has not aged well with such minimal context.
 
- @Mamaduka it looks that @jordesign added the `Needs Testing` but there is nothing really here to test, or if there is, we don't have clear testing instructions. Like @Successfulsebunya, I could only try a `npm run build` in a macOS and I can't get the heap error that @dmsnell is talking about. I think it's time to remove `Needs Testing` and add `Needs Test Info` and see if @dmsnell can provide some more info (or if he can decide to close this ticket all around)

**Comment by desrosj** [comment]
This was was fixed in Core `trunk` in [changeset 57250](https://core.trac.wordpress.org/changeset/57250) and this was opened up as a follow up. The [original thread in Slack can be found here](https://wordpress.slack.com/archives/C02RQBWTW/p1704724252878429?thread_ts=1704723980.921449&cid=C02RQBWTW) and has more details.

It's not known exactly why the the issue occurred, so I'm assuming the `needs testing` was added to try and figure out the actual cause of the issue.

In the https://github.com/wordpress/wordpress-develop workflows that test the build process, there is one that runs using a MacOS runner to ensure no issues while running the build scripts that would introduce problems for contributors. The suggestion here is to add a similar job to a workflow that `runs-on: macos-latest` (or whatever the current preferred runner is for MacOS) to try and catch issues like this before they're introduced.

**Comment by dmsnell** [comment]
Yeah there’s extremely low probability that attempting to build this once, or even ten times, will surface the issue. The issue itself, as @desrosj notes, is that currently we aren’t paying attention for these kinds of failures in the same way we are looking for similar failures for other platforms.

So unless this has happened already, we’re still just waiting for the next problem to show up and start crashing PRs in the `wordpress-develop` repo — some spooky action at a distance without any traceability back to Gutenberg.

The goal is to uncover failures on the `Gutenberg` side as soon as they happen and relieve people dealing with failures on `wordpress-develop` of the headaches from trying to figure out how their builds failed.

Expanding the memory was fixing a known and specific issue; this is about communicating when new specific issues arise.

**Comment by SirLouen** [comment]
The current situation is the following:

1. We don't have specific test cases yet. So basically, it's impossible to Test anything, What we need currently is Test Info that proves this casuistry.

2. After some research, I found [this commit](https://github.com/WordPress/gutenberg/commit/fc0573831cc7385cd7b65a2094cae4baac7b01be) that was within the time frame when the problem raised. I'm not sure if this could be affecting, but is the only commit that I can see that directly targets the build process. Significant changes in `lib/index.js`. Maybe they are irrelevant.

**Comment by SirLouen** [comment]
`Test Info`

This is only reproducible in a clean GB directory, after the first run, it will cache and not fail again. Build cache needs to be cleaned to consistently reproduce this issue.

It breaks under 3072 MB, and somewhere in between 4096 MB it starts moving, but it takes a year. 
I believe that the runner will work well around 5 GB, 8 GB is plenty.

For reproducing running default `npm run build` will do it, as a reference `NODE_OPTIONS="--max-old-space-size=3072" npm run build` to adjust. I believe this is ready for debugging what could be causing the issue, 

What I'm wondering how is possible that the amount of MacOS users running this are not finding this hindrance. By default, Node takes 2Gb so it should fail with all default setups. In Ubuntu (WSL2) this is not failing with defaults

**Comment by dmsnell** [comment]
@SirLouen I don’t think we should over-specify Node out-of-memory errors. I don’t even think we need to reproduce a failure.

What we want to ensure we do is _build_ Gutenberg in the Gutenberg repo on a macOS runner so that if and when it does start failing, we can prevent that failure from cascading remotely into the wordpress/WordPress-develop repo where it will be hard to diagnose and fix.

does that make sense? _any_ failure on macOS builds will currently start failing Core builds and that leaves no breadcrumbs for folks to follow to identify the cause.

**Comment by SirLouen** [comment]
I have not dug too far into this idea but why Gutenberg itself is not built in a monorepo fashion?

**Comment by dmsnell** [comment]
@SirLouen Gutenberg has been built here for many reasons.

Jobs in the CI pipeline for `WordPress/wordpress-develop` install the Gutenberg feature plugin in part to ensure that changes to Core don’t break Gutenberg.

There are a number of CI jobs that attempt to ensure that changes to Gutenberg don’t break core, just none that run on a macOS platform.

**Comment by SirLouen** [comment]
> [@SirLouen](https://github.com/SirLouen) Gutenberg has been built here for many reasons.
> There are a number of CI jobs that attempt to ensure that changes to Gutenberg don’t break core, just none that run on a macOS platform.

I'm unsure how to start working on this. 

Maybe it's a good time to consider transitioning from `webpack` into `vite` (or `esbuild`) ?

Related 
https://github.com/WordPress/gutenberg/issues/25077 
https://github.com/WordPress/gutenberg/discussions/36142

Perhaps a proposal must be suggested or a specific ticket for this?

**Comment by desrosj** [comment]
> Maybe it's a good time to consider transitioning from `webpack` into `vite` (or `esbuild`) ?

Transitioning build tools is a pretty substantial undertaking. I don't know that this issue is good justification for that.

The workflow in WordPress Develop that was added to try and detect failures around this sort of thing is this one: https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/reusable-test-gutenberg-build-process.yml.

To accomplish the goal of this ticket, we'd need a similar workflow in this repository that tests a similar process.

**Comment by SirLouen** [comment]
> To accomplish the goal of this ticket, we'd need a similar workflow in this repository that tests a similar process.

This looks a clearer target now for this ticket

> I don't know that this issue is good justification for that.

Apart from this, I think that this issue exposed the fact of the requirement to jump from defaults to 8GB 
https://github.com/WordPress/wordpress-develop/commit/71cb3f861573f065c6d7d7ef1975bf98532239b4

Maybe running on 8GB is irrelevant if the GH enterprise quotas are more than sufficient. 
But having to do this massive jump revealed the massive issue in optimization for the building process. It's not only happening on macos. I was able to also reproduce it on Ubuntu with WSL.

**Comment by dmsnell** [comment]
> Maybe it's a good time to consider transitioning from webpack into vite (or esbuild) ?

That would be a great question to ask in another ticket, as it’s not related to this issue other than a rare and incidental coincidence, that a memory limit was what crashed the build in macOS once. Tomorrow the issue could be entirely unrelated to memory, such as some filename macOS rejects, or a reliance on a system tool not available on macOS, or CLI argument syntax in the `gnu` release of a tool like `grep` where macOS and Linux distributions are typically divergent.

> I'm unsure how to start working on this.

The way forward I think is to work on the CI configs to also run a build of Gutenberg and WordPress on a macOS runner. Being able to build with a successful status code should pretty much resolve it.
